File: backend/app/schemas/hr.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import TYPE_CHECKING

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.approval_history import ApprovalHistoryEntry
from app.schemas.attachment import AttachmentResponse
from app.schemas.eligibility import EligibilityResponse
from app.schemas.payout import PayoutScheduleEntry

if TYPE_CHECKING:
    from app.models.claim import ClaimMaster


class ApproveRequest(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    approved_amount: Decimal = Field(gt=0)
    remarks: str | None = Field(default=None, max_length=1000)


class RejectRequest(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    remarks: str = Field(min_length=1, max_length=1000)


class SendBackRequest(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    remarks: str = Field(min_length=1, max_length=1000)


class HRClaimSummary(BaseModel):
    claim_id: int
    employee_id: str
    employee_name: str
    child_id: str
    child_name: str
    invoice_date: date
    invoice_number: str
    invoice_amount: Decimal
    claim_amount: Decimal
    claim_status: str
    comments: str | None
    submitted_date: datetime | None
    requires_documents: bool

    @classmethod
    def from_orm_model(
        cls, claim: ClaimMaster, *, employee_name: str, child_name: str, requires_documents: bool
    ) -> HRClaimSummary:
        return cls(
            claim_id=claim.ClaimID,
            employee_id=claim.EmployeeID,
            employee_name=employee_name,
            child_id=claim.ChildID,
            child_name=child_name,
            invoice_date=claim.InvoiceDate,
            invoice_number=claim.InvoiceNumber,
            invoice_amount=claim.InvoiceAmount,
            claim_amount=claim.ClaimAmount,
            claim_status=claim.ClaimStatus,
            comments=claim.Comments,
            submitted_date=claim.SubmittedDate,
            requires_documents=requires_documents,
        )


class HRClaimDetail(HRClaimSummary):
    eligibility: EligibilityResponse | None
    attachments: list[AttachmentResponse]
    approval_history: list[ApprovalHistoryEntry]
    payout_schedule: list[PayoutScheduleEntry]

File: frontend/src/api/hr.ts
import type { Attachment, PayoutScheduleEntry } from './claims'
import { apiClient } from '../lib/apiClient'
import type { EligibilitySummary } from './children'

export interface ApprovalHistoryEntry {
  approval_history_id: number
  action_by: string
  action_by_name: string
  action: 'Approved' | 'Rejected' | 'SentBack'
  previous_status: string
  new_status: string
  approved_amount: string | null
  remarks: string | null
  action_date: string
}

export interface HRClaimSummary {
  claim_id: number
  employee_id: string
  employee_name: string
  child_id: string
  child_name: string
  invoice_date: string
  invoice_number: string
  invoice_amount: string
  claim_amount: string
  institution_name: string | null
  from_date: string | null
  to_date: string | null
  claim_status: 'Draft' | 'Submitted' | 'HRReview' | 'Approved' | 'Rejected' | 'SentBack'
  comments: string | null
  submitted_date: string | null
  requires_documents: boolean
}

export interface HRClaimDetail extends HRClaimSummary {
  eligibility: EligibilitySummary | null
  attachments: Attachment[]
  approval_history: ApprovalHistoryEntry[]
  payout_schedule: PayoutScheduleEntry[]
}

export interface HRClaimListFilters {
  status?: string
  employee_id?: string
  child_id?: string
  date_from?: string
  date_to?: string
}

export async function listHRClaims(filters: HRClaimListFilters = {}): Promise<HRClaimSummary[]> {
  const { data } = await apiClient.get<HRClaimSummary[]>('/hr/claims', { params: filters })
  return data
}

export async function getHRClaimDetail(claimId: number): Promise<HRClaimDetail> {
  const { data } = await apiClient.get<HRClaimDetail>(`/hr/claims/${claimId}`)
  return data
}

export async function approveClaim(
  claimId: number,
  approvedAmount: string,
  remarks?: string,
): Promise<HRClaimDetail> {
  const { data } = await apiClient.post<HRClaimDetail>(`/hr/claims/${claimId}/approve`, {
    approved_amount: approvedAmount,
    remarks: remarks || undefined,
  })
  return data
}

export async function rejectClaim(claimId: number, remarks: string): Promise<HRClaimDetail> {
  const { data } = await apiClient.post<HRClaimDetail>(`/hr/claims/${claimId}/reject`, {
    remarks,
  })
  return data
}

export async function sendBackClaim(claimId: number, remarks: string): Promise<HRClaimDetail> {
  const { data } = await apiClient.post<HRClaimDetail>(`/hr/claims/${claimId}/send-back`, {
    remarks,
  })
  return data
}

export async function downloadHRAttachment(
  claimId: number,
  attachmentId: number,
  filename: string,
): Promise<void> {
  const response = await apiClient.get(
    `/hr/claims/${claimId}/attachments/${attachmentId}/download`,
    { responseType: 'blob' },
  )
  const url = window.URL.createObjectURL(response.data as Blob)
  const link = document.createElement('a')
  link.href = url
  link.download = filename
  document.body.appendChild(link)
  link.click()
  link.remove()
  window.URL.revokeObjectURL(url)
}
